### PR TITLE
feat(aiplatform): top level folder with initial PaLM API samples

### PR DESCRIPTION
**Pull request by ghchinoy**
## Description

Adds aiplatform top level folder with initial PaLM API samples

## Checklist
- [x] I have followed [Contributing Guidelines from CONTRIBUTING.MD](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md)
- [x] **Tests** pass:   `go test -v ./..` (see [Testing](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#testing))
- [x] **Code formatted**:   `gofmt` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [x] **Vetting** pass:   `go vet` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [x] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: aiplatform/text-embeddings/embeddings_test.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"bytes"
+	"strings"
+	"testing"
+)
+
+func TestGenerateEmbeddings(t *testing.T) {
+
+	prompt := "hello, say something nice."
+	projectID := "PROJECT_ID"
```
Comment (telpirion) [inline aiplatform/text-embeddings/embeddings_test.go:26]: issue: use `testutil.SystemTest(t)` here to get the `TestContext`. You can access the project ID from the `TestContext`.

File: aiplatform/text-predict/text.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	"io"
+	"log"
+
+	aiplatform "cloud.google.com/go/aiplatform/apiv1beta1"
+	"cloud.google.com/go/aiplatform/apiv1beta1/aiplatformpb"
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// [START aiplatform_text_predictions]
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:30]: issue: move region tags before `import` statement.

File: aiplatform/text-predict/text.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+func TextPredict(w io.Writer, prompt, projectID, location, publisher, model string, parameters map[string]interface{}) error {
+	ctx := context.Background()
+
+	apiEndpoint := fmt.Sprintf("%s-aiplatform.googleapis.com:443", location)
+	log.Printf("apiendpoint: %s", apiEndpoint)
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:36]: issue: remove `log.Printf()` here.

File: aiplatform/text-predict/text.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	c, err := aiplatform.NewPredictionClient(ctx, option.WithEndpoint(apiEndpoint))
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:38]: issue: replace `c` with `client`.

File: aiplatform/text-predict/text.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		log.Printf("unable to create prediction client: %v", err)
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:40]: issue: replace `log.Printf()` with `fmt.Fprintf()`.

File: aiplatform/text-predict/text.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		return err
+	}
+	defer c.Close()
+
+	// PredictRequest requires an Endpoint, Instances, and Parameters
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:45]: issue: see previous about capitalization.

File: aiplatform/text-predict/text.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	// Endpoint
+	base := fmt.Sprintf("projects/%s/locations/%s/publishers/%s/models", projectID, location, publisher)
+	url := fmt.Sprintf("%s/%s", base, model)
+
+	// Instances: the prompt
+	promptValue, err := structpb.NewValue(map[string]interface{}{
+		"prompt": prompt,
+	})
+	if err != nil {
+		log.Printf("unable to convert prompt to Value: %v", err)
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:55]: issue: use `fmt.Fprintf()` and return error here.

File: aiplatform/text-predict/text.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	// Parameters: the model configuration parameters
+	parametersValue, err := structpb.NewValue(parameters)
+	if err != nil {
+		log.Printf("unable to convert parameters to Value: %v", err)
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:61]: issue: same as previous (use `fmt.Fprintf()` and return error).

File: aiplatform/text-predict/text.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	// PredictRequest
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:64]: issue: provide more explanatory comment or cut.

File: aiplatform/text-predict/text.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	req := &aiplatformpb.PredictRequest{
+		Endpoint:   url,
+		Instances:  []*structpb.Value{promptValue},
+		Parameters: parametersValue,
+	}
+	log.Printf("PredictRequest.Endpoint:   %v", req.GetEndpoint())
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:70]: issue: cut these three lines.

File: aiplatform/text-predict/text.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+	log.Printf("PredictRequest.Instances:  %v", req.GetInstances())
+	log.Printf("PredictRequest.Parameters: %v", req.GetParameters())
+
+	// PredictResponse
+	resp, err := c.Predict(ctx, req)
+	if err != nil {
+		log.Printf("error in prediction: %v", err)
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:77]: issue: use `fmt.Fprintf()` here.

File: aiplatform/text-predict/text.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+		return err
+	}
+
+	jsonData, err := json.MarshalIndent(resp, "", " ")
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:81]: issue/question: can you access the fields of the `PredictResponse` here instead?

File: aiplatform/text-predict/text.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+	log.Printf("%s\n", jsonData)
```
Comment (telpirion) [inline aiplatform/text-predict/text.go:82]: issue: cut this line.

File: aiplatform/text-predict/text_test.go [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+	//"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+)
+
+func TestTextPredict(t *testing.T) {
+	//testutil.SystemTest(t)
```
Comment (telpirion) [inline aiplatform/text-predict/text_test.go:25]: issue:  uncomment.

File: aiplatform/text-predict/text_test.go [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snippets
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+	//"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
```
Comment (telpirion) [inline aiplatform/text-predict/text_test.go:21]: issue: uncomment.

File: aiplatform/text-predict/text_test.go [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+func TestTextPredict(t *testing.T) {
+	//testutil.SystemTest(t)
+
+	prompt := "hello, say something nice."
+	projectID := "PROJECT_ID"
```
Comment (telpirion) [inline aiplatform/text-predict/text_test.go:28]: issue: access the ProjectID field from the object returned from `testutil.SystemTest(t)`.

**Review by telpirion — CHANGES_REQUESTED** [review]

File: aiplatform/text-embeddings/embeddings.go [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+	if err != nil {
+		log.Printf("unable to convert prompt to Value: %v", err)
+	}
+
+	// PredictRequest
+	req := &aiplatformpb.PredictRequest{
+		Endpoint:  url,
+		Instances: []*structpb.Value{promptValue},
+	}
+	log.Printf("PredictRequest.Endpoint:   %v", req.GetEndpoint())
```
[the remark this answers is not in the document]
Comment (ghchinoy) [reply aiplatform/text-embeddings/embeddings.go:63]: They were, yes -  will cut.

File: aiplatform/text-embeddings/embeddings.go [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+	// Instances: the prompt
+	promptValue, err := structpb.NewValue(map[string]interface{}{
+		"content": prompt,
+	})
+	if err != nil {
+		log.Printf("unable to convert prompt to Value: %v", err)
+	}
+
+	// PredictRequest
+	req := &aiplatformpb.PredictRequest{
```
[the remark this answers is not in the document]
Comment (ghchinoy) [reply aiplatform/text-embeddings/embeddings.go:59]: In the test, the Parameters are assigned and govern the text model attributes (temperature, max output tokens, top P, top K)

File: aiplatform/text-embeddings/embeddings.go [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+		Endpoint:  url,
+		Instances: []*structpb.Value{promptValue},
+	}
+	log.Printf("PredictRequest.Endpoint:   %v", req.GetEndpoint())
+	log.Printf("PredictRequest.Instances:  %v", req.GetInstances())
+	log.Printf("PredictRequest.Parameters: %v", req.GetParameters())
+
+	// PredictResponse
```
[the remark this answers is not in the document]
Comment (ghchinoy) [reply aiplatform/text-embeddings/embeddings.go:67]: Will add more information

File: aiplatform/text-embeddings/embeddings.go [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+	resp, err := c.Predict(ctx, req)
+	if err != nil {
+		log.Printf("error in prediction: %v", err)
+		return err
+	}
+	jsonData, err := json.MarshalIndent(resp, "", " ")
```
[the remark this answers is not in the document]
Comment (ghchinoy) [reply aiplatform/text-embeddings/embeddings.go:73]: Yes, another debug artifact to be removed. thank you!

File: aiplatform/text-embeddings/embeddings_test.go [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"bytes"
+	"strings"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+)
+
+func TestGenerateEmbeddings(t *testing.T) {
+	testutil.SystemTest(t)
```
Comment (telpirion) [inline aiplatform/text-embeddings/embeddings_test.go:26]: nit: `SystemTest` returns an object that you can keep a reference for later (e.g. line 29).

File: aiplatform/text-predict/text_test.go [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	"bytes"
+	"strings"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+)
+
+func TestTextPredict(t *testing.T) {
+	testutil.SystemTest(t)
```
Comment (telpirion) [inline aiplatform/text-predict/text_test.go:26]: Same as previous: you can capture a reference to the return value of `SystemTest` and then dot into it below.

**Review by telpirion — APPROVED** [review]

**Pull request merged**